### PR TITLE
added support for swipeWidget on swipe in listView Item

### PR DESCRIPTION
**Pull request by TheNoumanDev**
ticket: https://github.com/EnsembleUI/ensemble/issues/1646

Example EDL:
```yaml
View:
  header:
    title: "ListView: Slidable Example"
  onLoad:
    invokeAPI:
      name: getPeople

  body:
    Column:
      children:
        - ListView:
            id: listView
            styles: { expanded: true }
            item-template:
              data: ${getPeople.body.users}
              name: users
              template:
                Slidable:
                  styles:
                    direction: horizontal
                    closeOnScroll: true
                    DragStartBehavior: start
                    useTextDirection: true
                  startContent:
                    options:
                      extentRatio: 0.5
                      openThreshold: 0.25
                      motion: behind
                    children:
                      - icon: star
                        backgroundColor: red
                        foregroundColor: green
                        borderRadius: 20
                      - label: Cancel
                        onPressed:
                          showToast:
                            message: User deleted
                  endContent:
                    children:
                      - icon: star
                        label: Favorite
                        backgroundColor: yellow
                        onPressed:
                          showToast:
                            message: Added to favorites
                  child:
                    MyRow:
                      inputs:
                        p: ${users}

MyRow:
  inputs:
    - p
  body:
    Column:
      children:
        - Row:
            styles:
              gap: 12
              crossAxis: center
              padding: 8 0
            children:
              - Image:
                  styles: { width: 50, height: 50, borderRadius: 50 }
                  source: ${p.image}
              - Column:
                  styles: { gap: 2 }
                  children:
                    - Text:
                        styles: { fontWeight: bold }
                        text: ${p.firstName} ${p.lastName}
                    - Text:
                        text: ${p.email}
              - Spacer
              - Icon:
                  icon: chevron_right
                  styles: { size: 20, color: black }

API:
  getPeople:
    uri: https://dummyjson.com/users
    method: GET
```

will add schema PR, docs and kitchen SInk once PR gets approved before merging it,

### CODE REVIEW

**Comment by vusters** [comment]
1. Your swipeWidget will not work. The reason it seems to work is because you simply print out to console. Each ListView's item (i.e. each children and each item template) has to be wrapped inside a Swipeable, otherwise how would you render say a Delete button for each and can tell which item you actually want to delete on? A simpler way to think about this is how would I display a unique piece of text differently for each item's swipe out section? You can't.

2. Never write code if you don't have to. It's a liability. Use this package that has 5.5K stars and 100% - mean well tested https://pub.dev/packages/flutter_slidable

3. What you want to implement is probably a generic **Card** widget that can swipe left/right/up/down to expose more content. To keep it simple for now let's just call it "Slidable" and follow the package structure (it's a decent structure)
```
ListView:
  itemTemplate:
    data: ...
    name: item
    template:
      Slidable:
        body: <your_widget_to_display_as_default_state>
        startContent: <widget_to_display_on_swipe_right>
        endContent: <widget_to_display_on_swipe_left>
```

Use start/end vs left/right to account for RTL languages.

**Comment by kmahmood74** [comment]
@TheNoumanDev please make the changes that Vu suggested

**Comment by TheNoumanDev** [comment]
> @TheNoumanDev please make the changes that Vu suggested

On it.

**Comment by kmahmood74** [comment]
@TheNoumanDev any update?

**Comment by TheNoumanDev** [comment]
@kmahmood74 Sorry for delay, was stuck with KPN task.
@vusters PTAL. Thanks

File: modules/ensemble/lib/layout/list_view.dart [hunk 1]
```diff
@@ -133,22 +124,16 @@ class ListViewController extends BoxLayoutController {
...
-  EnsembleAction? onSwipe;
-
   void _bind(ListViewState state) {
     widgetState = state;
   }
 }
 
 class ListViewState extends EWidgetState<ListView>
-    with TickerProviderStateMixin, TemplatedWidgetState, HasChildren<ListView> {
+    with TemplatedWidgetState, HasChildren<ListView> {
```
Comment (vusters) [inline modules/ensemble/lib/layout/list_view.dart]: Just to confirm, there is no code changes in list_view right? This is just a revert to the original version?

File: modules/ensemble/lib/widget/slidable.dart [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+  }
+
+  @override
+  Map<String, Function> methods() {
+    return { };
+  }
+}
+
+class ActionPaneOptionsComposite {
+  ActionPaneOptionsComposite() {
```
Comment (vusters) [inline modules/ensemble/lib/widget/slidable.dart:64]: dont' repeat the code in here.

File: modules/ensemble/lib/widget/slidable.dart [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  @override
+  EnsembleSlidableController get controller => _controller;
+
+  @override
+  State<StatefulWidget> createState() => EnsembleSlidableState();
+
+  @override
+  Map<String, Function> getters() {
+    return {
+      'startContent': () => _controller.startContent,
```
Comment (vusters) [inline modules/ensemble/lib/widget/slidable.dart:37]: prob should mention this before - we need more descriptive name: startDrawer/endDrawer? @amin-nas

File: modules/ensemble/lib/widget/slidable.dart [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    };
+  }
+
+  @override
+  Map<String, Function> methods() {
+    return { };
+  }
+}
+
+class ActionPaneOptionsComposite {
```
Comment (vusters) [inline modules/ensemble/lib/widget/slidable.dart:63]: This has to extend a base composite for binding to work.

File: modules/ensemble/lib/widget/slidable.dart [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+        color: iconModel.color,
+      ).icon;
+    }
+
+    return iconframework.Icon(value).icon;
+  }
+
+  Widget _buildChildWidget(BuildContext context, dynamic widgetDefinition) {
+    if (widgetDefinition == null) {
+      return Container();
```
Comment (vusters) [inline modules/ensemble/lib/widget/slidable.dart:216]: use ShrinkBox.empty()

File: modules/ensemble/lib/layout/list_view.dart (hunk 1 above)
Comment (TheNoumanDev) [reply modules/ensemble/lib/layout/list_view.dart]: Yes that is outdated, there is no changes in listView.

File: modules/ensemble/lib/widget/slidable.dart (hunk 5 above)
Comment (TheNoumanDev) [reply modules/ensemble/lib/widget/slidable.dart]: Updated!

File: modules/ensemble/lib/widget/slidable.dart (hunk 3 above)
Comment (TheNoumanDev) [reply modules/ensemble/lib/widget/slidable.dart]: Updated it to `startDrawer/endDrawer` for now, do let me if something better came up😅

File: modules/ensemble/lib/widget/slidable.dart [hunk 6]
```diff
@@ -136,61 +97,72 @@ class EnsembleSlidableState extends EWidgetState<EnsembleSlidable> {
...
-        return const SizedBox.shrink(); // Return an empty widget if both are missing
+        return const SizedBox
+            .shrink(); // Return an empty widget if both are missing
       }
 
       return SlidableAction(
         onPressed: (context) {
           if (action['onPressed'] != null) {
-            final onPressedAction = EnsembleAction.from(action['onPressed'], initiator: widget);
+            final onPressedAction =
```
Comment (vusters) [inline modules/ensemble/lib/widget/slidable.dart:143]: This certainly can return null which will throw an except below. Check for it and never never ever use force exclamation like "onPressAction!". If you have to do that, something is wrong.

**Review by vusters — APPROVED** [review]
You are good to check if AFTER addressing the stated.

File: modules/ensemble/lib/widget/slidable.dart (hunk 6 above)
Comment (TheNoumanDev) [reply modules/ensemble/lib/widget/slidable.dart:143]: yup got the point, thanks, and will take care in future as well, but the same has been implemented at every action, maybe we need to keep updating such mistake while fixing other stuff.

**Pull request merged**